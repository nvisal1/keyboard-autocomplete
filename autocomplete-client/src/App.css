.Autocomplete-client {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: #6C6C6C;
}

