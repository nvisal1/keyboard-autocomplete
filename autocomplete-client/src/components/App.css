.Autocomplete-client {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: var(--asymmetrik-light-grey);
}

.Autocomplete-client__Navbar-container {
  position: absolute;
  top: 0;
  width: 100%;
}

.Autocomplete-client__Input-container {
  max-width: 1440px;
  padding-left: 40px;
  padding-right: 40px;
  height: 30%;
  margin: auto;
  position: absolute;
  top: 0; left: 0; bottom: 0; right: 0;
}

.Autocomplete-client__train-form-container {
  max-width: 1440px;
  padding-left: 40px;
  padding-right: 40px;
  height: 50%;
  margin: auto;
  position: absolute;
  top: 0; left: 0; bottom: 0; right: 0;
}

.Autocomplete-client__Farley-container {
  position: absolute;
  bottom: 70px;
}

.Autocomplete-client__Footer-container {
  position: absolute;
  bottom: 0;
  width: 100%;
  z-index: -1;
}
