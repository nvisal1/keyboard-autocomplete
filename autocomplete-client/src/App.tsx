import React from 'react';
import './App.css';
import Navbar from './components/navbar/navbar';
import Footer from './components/footer/footer';
import Input from './components/input/input';
import Farley from './components/Farley/Farley';
import server from './shared/server';
import { Candidate } from './shared/types/Candidate';

interface AppState {
  candidates: Candidate[];
}

class App extends React.Component<any, AppState> {

  constructor(props: any) {
    super(props)
    this.state = {
        candidates: []
    }
  }

  render() { 
    return (
      <div className='Autocomplete-client'>
        <div className='Autocomplete-client__Navbar-container'>
          <Navbar></Navbar>
        </div>

        <div className='Autocomplete-client__Input-container '>
          <Input handleInput = { this.handleInput }></Input>
        </div>

        <div className='Autocomplete-client__Farley-container'>
          <Farley candidates={ this.state.candidates }></Farley>
        </div>

        <div className='Autocomplete-client__Footer-container'>
          <Footer></Footer>
        </div>
      </div>
    );
  }

  handleInput = async(text: string): Promise<void> => {
    try {
      const tokens = text.split(' '); 
      const currentFragment = tokens[tokens.length - 1];
      const response = await server().get(`/candidates?text=${ currentFragment }`);
      const candidates: Candidate[] = response.data;
      this.setState({ candidates });
    } catch (error) {
      this.setState({ candidates: [] });
    }
  }
}
export default App;
