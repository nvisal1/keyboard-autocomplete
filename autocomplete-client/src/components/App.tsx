import React from 'react';
import './App.css';
import Navbar from './navbar/navbar';
import Footer from './footer/footer';
import Input from './input/input';
import Farley from './Farley/Farley';
import { Candidate } from '../shared/types/Candidate';
import TrainForm from './train-form/train-form';
import { AUTOCOMPLETE_PROVIDER } from '../drivers';

enum Modes {
  SEARCH = 'Search',
  TRAIN = 'Train',
  BOTH = 'Search & Train',
}

const modes = [ Modes.SEARCH, Modes.TRAIN, Modes.BOTH ];

interface AppState {
  candidates: Candidate[];
  farleySteps: number;
  text: string;
  mode: string;
  error: {
    message: string,
    isError: boolean,
  }
}

const MAX_FARLEY_STEPS = window.innerWidth - 500;

class App extends React.Component<{}, AppState> {

  constructor(props: any) {
    super(props)
    this.state = {
        candidates: [],
        farleySteps: 0,
        text: '',
        mode: 'Search',
        error: {
          message: '',
          isError: false,
        },
    }
  }

  render() { 
    return (
      <div className='Autocomplete-client'>
        <div className='Autocomplete-client__Navbar-container'>
          <Navbar 
            selectedOption={ this.state.mode }
            options={ modes }
            handleClick={ this.handleModeChange }
          ></Navbar>
        </div>

        { this.renderInput() }  

        <div 
          className='Autocomplete-client__Farley-container'
          style={ { marginLeft: this.state.farleySteps + 'px' } }>
          <Farley 
            errorMessage={ this.state.error.isError ? this.state.error.message : undefined }
            candidates={ this.state.candidates }
            handleClick={ this.handleCandidateSelection }
          ></Farley>
        </div>

        <div className='Autocomplete-client__Footer-container'>
          <Footer></Footer>
        </div>
      </div>
    );
  }

  renderInput(): JSX.Element {
    if (this.state.mode === Modes.TRAIN) {
      return (
        <div className='Autocomplete-client__train-form-container'>
          <TrainForm
            handleSubmit={ this.handleTrainSubmission }
          ></TrainForm>
        </div>
      );
    } else {
      return (
        <div className='Autocomplete-client__Input-container'>
          <Input 
            handleInput={ this.handleInputChange }
            text={ this.state.text }
          ></Input>
        </div>
      );
    }
  }

  handleModeChange = (mode: string) => {
    this.setState({ mode });
  }

  handleInputChange = async(text: string): Promise<void> => {
    this.setState({ text });
    
    const currentFragment = this.getCurrentWordFragment(text);

    const candidates = await this.getCandidates(currentFragment);

    if (candidates) {
      this.setState({ 
        candidates, 
        farleySteps: this.calculateFarleySteps(text),
      });
    }
  }

  handleTrainSubmission = (passage: string): void => {
    try {
      if (passage) {
        AUTOCOMPLETE_PROVIDER.train(passage);
      }
    } catch (error) {
      this.setState({ error: { message: error.message, isError: true }});
    }
  }

  handleCandidateSelection = (word: string): void => {
    const updatedInputText = this.generateUpdatedInputText(word);
    this.setState({ text: updatedInputText, candidates: [] });
  }

  private getCurrentWordFragment(text: string): string {
    const tokens = text.split(' '); 
    const currentFragment = tokens[tokens.length - 1];
    return currentFragment;
  }

  private async getCandidates(currentFragment: string): Promise<Candidate[] | undefined> {
    try {
      if (currentFragment !== '') {
        const candidates = await AUTOCOMPLETE_PROVIDER.getWords(currentFragment);
        return candidates;
      }
      return [];
    } catch (error) {
      this.setState({ error: { message: error.message, isError: true }});
    }
  }

  private calculateFarleySteps(text: string) {
    const newFarleyStep = text.length * 15;

    if (MAX_FARLEY_STEPS > newFarleyStep) {
      return newFarleyStep;
    }

    return this.state.farleySteps
  }

  private generateUpdatedInputText(word: string) {
    const tokens = this.state.text.split(' '); 

    tokens[tokens.length - 1] = word;

    let updatedInputText = '';
    tokens.forEach((token: string) => {
      updatedInputText += token + ' ';
    });

    return updatedInputText;
  }

}
export default App;
