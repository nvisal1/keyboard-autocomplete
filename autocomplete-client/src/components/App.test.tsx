import React from 'react';
import renderer from 'react-test-renderer';

import App from './App';

describe('When the App component is created', () => {
    it('should render correctly', () => {
        const currentSnapshot = renderer.create(<App></App>).toJSON();

        expect(currentSnapshot).toMatchSnapshot();
    });
});
