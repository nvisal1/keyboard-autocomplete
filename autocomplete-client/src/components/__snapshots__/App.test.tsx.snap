// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When the App component is created should render correctly 1`] = `
<div
  className="Autocomplete-client"
>
  <div
    className="Autocomplete-client__Navbar-container"
  >
    <div
      className="navbar"
    >
      <div
        className="navbar__options-container"
      >
        <div
          className="option-container__text"
          onClick={[Function]}
          style={
            Object {
              "color": "",
            }
          }
        >
          Search
        </div>
        <div
          className="option-container__text"
          onClick={[Function]}
          style={
            Object {
              "color": "",
            }
          }
        >
          Train
        </div>
        <div
          className="option-container__text"
          onClick={[Function]}
          style={
            Object {
              "color": "",
            }
          }
        >
          Search & Train
        </div>
      </div>
    </div>
  </div>
  <div
    className="Autocomplete-client__Input-container"
  >
    <div
      className="input-container"
    >
      <span
        className="input__icon-cntainer"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
      <input
        className="input-container__input"
        onChange={[Function]}
        value=""
      />
    </div>
  </div>
  <div
    className="Autocomplete-client__Farley-container"
    style={
      Object {
        "marginLeft": "0px",
      }
    }
  >
    <div
      className="farley"
    >
      <img
        alt="Farley"
        className="farley__image-container"
        src="Farley.png"
      />
      <div
        className="farley__speech-bubble-container"
      >
        <div
          className="speech-bubble"
        >
          <div
            className="speech-bubble__text-container"
          >
            <div
              className="speech-bubble__pointer-container"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="Autocomplete-client__Footer-container"
  >
    <div
      className="footer"
    />
  </div>
</div>
`;
